### ISSUE TITLE
Mapbox crashes on setCamera

### ISSUE THREAD

**Issue by SamuelBrucksch** [mapbox/mapbox-maps-android#649]
## Environment

- Android OS version: 11
- Devices affected: Pixel3
- Maps SDK Version: 10 RC8

Hi,

i get this stacktrace reported on a crash that randomly appears on android auto:
```
java.lang.Error: std::exception
        at com.mapbox.maps.CameraManager.setCamera(CameraManager.java:-2)
        at com.mapbox.maps.NativeMapImpl.setCamera(NativeMapImpl.kt:35)
        at com.mapbox.maps.MapboxMap$setCamera$1.invoke(MapboxMap.kt:280)
        at com.mapbox.maps.MapboxMap$setCamera$1.invoke(MapboxMap.kt:37)
        at com.mapbox.maps.UtilsKt.call(Utils.kt:9)
        at com.mapbox.maps.MapboxMap.setCamera(MapboxMap.kt:280)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.performMapJump(CameraAnimationsPluginImpl.kt:185)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.access$performMapJump(CameraAnimationsPluginImpl.kt:36)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$commitChangesRunnable$1.run(CameraAnimationsPluginImpl.kt:62)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7664)
        at java.lang.reflect.Method.invoke(Method.java:-2)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```
Unfortunately it does not contain any information, which part of my code was executed before this. I would assume it was caused by a call of setCamera but later in the stack trace i also see the Camera Animator Plugin. So i'm not really sure where to look for.

Is there any way to find the problem here? What could typically be the problem here?

**Comment by kiryldz** [comment]
Hi @SamuelBrucksch!
Could you please as maximum info as possible:
 - how often does this happen?
 - are there any preconditions, does it happen during app minimize / maximize or during regular map update with no visible preconditions?
 - ideally it would be great if you could provide code snippet you're using so that we could try reproducing it on our end

**Comment by SamuelBrucksch** [comment]
Hi,

this happens randomly once or twice a day. I use a MapSurface and get the MapboxMap from `mapSurface.getMapboxMap()` like this:
```
                Surface s = surfaceContainer.getSurface();
                mapSurface = new MapSurface(mCarContext, s, options);
                mapSurface.setMaximumFps(30);
                mapboxMap = mapSurface.getMapboxMap();
```
The surface comes from `SurfaceCallback` -> `onSurfaceAvailable()`

This happens while the map is displayed. I can not share my whole code, so if you could tell me which specific parts could cause this i can maybe share some snippets. In general i only use `mapboxMap.setCamera()` and `cameraAnimations.easeTo`.

Would it work to surround the car animator calls with a try catch to identify the exact position or would that not throw an exception as some parts are called in background threads?

**Comment by kiryldz** [comment]
> Hi,
> 
> this happens randomly once or twice a day (per user). I use a MapSurface and get the MapboxMap from `mapSurface.getMapboxMap()` like this:
> 
> ```
>                 Surface s = surfaceContainer.getSurface();
>                 mapSurface = new MapSurface(mCarContext, s, options);
>                 mapSurface.setMaximumFps(30);
>                 mapboxMap = mapSurface.getMapboxMap();
> ```
> 
> The surface comes from `SurfaceCallback` -> `onSurfaceAvailable()`
> 
> This happens while the map is displayed. I can not share my whole code, so if you could tell me which specific parts could cause this i can maybe share some snippets. In general i only use `mapboxMap.setCamera()` and `cameraAnimations.easeTo`.
> 
> Would it work to surround the car animator calls with a try catch to identify the exact position or would that not throw an exception as some parts are called in background threads? Because if the exception would reach my code i guess i would see it in the stacktrace.
> 
> The problem here is that i don't know where it crashes, so i can not check the preconditions.

It crashes somewhere in native code.
Is it possible for you to check logcat before exception is thrown and see if you could see anything suspicious there? Is it possible that you're calling `mapboxMap.setCamera()` while `cameraAnimations.easeTo` is already running?

**Comment by SamuelBrucksch** [comment]
Hi,

as this is running on android auto i can not see the logcat logs. The exception was reported via Bugsnag, so the exception above is the only thing i can see. 

setCamera should normally not be called during navigation. Actually this is the function, that normally should be called during navigation:
```
    public void animateMapTo(Point mapCenter, double bearing, double pitch, double zoom) {
        if (mapAnimator != null) {
            mapAnimator.cancel();
            mapAnimator = null;
        }

        if (userTouchInProgress || isRoutePreviewEnabled())
            return;

        currentCamera.center(mapCenter).bearing(bearing).zoom(zoom).pitch(pitch);

        if (cameraAnimations == null) return;
        mapAnimator = cameraAnimations.easeTo(currentCamera.build(), new MapAnimationOptions.Builder().duration(1000).interpolator(new LinearInterpolator()).build());
    }
```

or this if the touchscreen is tapped and the buttons show up:
```
        @Override
        public void onVisibleAreaChanged(@NonNull Rect visibleArea) {
            synchronized (MapRenderer.this) {
                // Log.i(TAG, "onVisibleAreaChanged W: " + visibleArea.toString());
                MapRenderer.this.edgeInsets = new EdgeInsets(minRequiredTop + 10, visibleArea.left + 10, 10, 10);

                if (!isReady) return;

                currentCamera.padding(edgeInsets);
                cameraAnimations.easeTo(currentCamera.build(), new MapAnimationOptions.Builder().duration(500).build());
            }
        }
```

`currentCamera` comes from
```
    private final CameraOptions.Builder currentCamera = new CameraOptions.Builder().zoom(14d);
```
which i store to be able to immediately show the correct camera when we come back from background.

**Comment by SamuelBrucksch** [comment]
Now that i see those code snippets... Could it happen, that `easeTo` crashes, if another animation is already running? Should i store the current animation and always cancel it before easeTo is called?

**Comment by kiryldz** [comment]
> Now that i see those code snippets... Could it happen, that `easeTo` crashes, if another animation is already running? Should i store the current animation and always cancel it before easeTo is called? However when i just tried this, it doesnt crash, so it should work i think...

It should perform all the cancelling automatically inside animation plugin.
However if your code is designed in a way that could call `mapboxMap.setCamera` directly - it will not go thru camera animation plugin. 

If you have capacity you could try either replace `mapboxMap.setCamera` calls with `easeTo` with `duration = 0` or try calling `mapView.camera.cancelAllAnimators(listOf())` before `mapboxMap.setCamera` calls. I'm not sure though it will fix an issue but we need more data in order to understand root cause.

**Comment by SamuelBrucksch** [comment]
Hi @kiryldz ,

i checked our code again and setCamera is not called during navigation. So it really is only the `easeTo` that causes the crash.

I also checked our logs again and that error occured first on August 19th where we used RC5 for the first time. I need to additionally check, if we changed any code with that release, but it looks like we did not see the error before.

**Comment by kiryldz** [comment]
@SamuelBrucksch I'm not sure this error was introduced between rc3 and rc4, it happens __very__ rarely for quite a time already but we can't get exact pre-conditions as well as reproduce it somehow.
However we may have a potential solution, we will try to bring in the fix before our next release happening this week.

**Comment by SamuelBrucksch** [comment]
Great, will happily test it once it is out.

**Comment by kiryldz** [comment]
@SamuelBrucksch I would really appreciate your help if you could test following setup. This setup does not have https://github.com/mapbox/mapbox-maps-android/pull/652 "fix" but tries to fix the root cause. However taking into account that we could not replicate this crash (yet) on our side it would be nice if you could check it on your end and see if this crash is fixed.

Add this block to high-level project level `build.gradle` file:

```
maven {
    url 'https://api.mapbox.com/downloads/v2/snapshots/maven'
    authentication {
        basic(BasicAuthentication)
    }
    credentials {
        username = "mapbox"
        password = <SAME_TOKEN_YOU_USE_ALREADY_GOES_HERE>
    }
}
```

Update your module level `build.gradle` file to make use of force dependency:

```
    // Mapbox
    implementation 'com.mapbox.maps:android:10.0.0-rc.8'
    configurations.all {
        resolutionStrategy {
            force 'com.mapbox.maps:android-core:10.0.0-rc.8-563f20769bfd090f0aaf065a59e258519d951c6a-SNAPSHOT'
        }
    }
```

cc @ank27 @alexshalamov

**Comment by SamuelBrucksch** [comment]
Ok I'll just let it running tonight and filter in Logcat for `java.lang.Error` and check again tomorrow if i see anything.

**Comment by SamuelBrucksch** [comment]
Hi,

it was running for quite some time without crash but now i just got another one:
```
2021-09-22 10:33:42.232 E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.iternio.abrpapp.auto, PID: 24352
    java.lang.Error: std::exception
        at com.mapbox.maps.CameraManager.setCamera(Native Method)
        at com.mapbox.maps.NativeMapImpl.setCamera(NativeMapImpl.kt:35)
        at com.mapbox.maps.MapboxMap$setCamera$1.invoke(MapboxMap.kt:280)
        at com.mapbox.maps.MapboxMap$setCamera$1.invoke(MapboxMap.kt:37)
        at com.mapbox.maps.UtilsKt.call(Utils.kt:9)
        at com.mapbox.maps.MapboxMap.setCamera(MapboxMap.kt:280)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.performMapJump(CameraAnimationsPluginImpl.kt:185)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl.access$performMapJump(CameraAnimationsPluginImpl.kt:36)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$commitChangesRunnable$1.run(CameraAnimationsPluginImpl.kt:62)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6626)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:811)
```

So if that snapshot version already contained a fix it did not seem to help or there is an additional one.

**Comment by kiryldz** [comment]
@SamuelBrucksch thanks for checking it! Really appreciate it.
In that case we'll continue finding root cause but this fix https://github.com/mapbox/mapbox-maps-android/pull/652 will be landing in release that should happen today and this should solve at least an issue with crashing 😄

**Comment by SamuelBrucksch** [comment]
Great. Should we leave this open so i can provide the camera options when a crash is logged? Maybe that helps in writing tests to identify the root cause.

**Comment by kiryldz** [comment]
No, when you'll verify that no crashes will appear with latest release, we could close this one.
We have internal tickets to track finding the root cause.

**Comment by SamuelBrucksch** [comment]
I could see a message in the log:
```
2021-09-22 20:04:53.125 E/Mbgl-CameraManager: Error while setting camera options : std::exception CameraOptions = [center: Point{type=Point, bbox=null, coordinates=[8.605256001252144, 49.26057483076021]}, padding: [top: 45.0, left: 346.0, bottom: 10.0, right: 10.0], anchor: , zoom: 13.25, bearing: 22.88029118538884, pitch: 65.0]
```

From what i see in the log the app did not restart. So it seems to work.

**Comment by pengdev** [comment]
Closing as the fix is included in v10.0.0-rc.9 release.
